[IR] Allow fast math flags on fptrunc and fpext

john-brawn-arm opened the pull request:
This consists of:
 * Make these instructions part of FPMathOperator.
 * Adjust bitcode/ir readers/writers to expect fast math flags on these instructions.
 * Make IRBuilder set the fast math flags on these instructions.
 * Update langref and release notes.
 * Update a bunch of tests. Some of these are due to InstCombineCasts incorrectly adding fast math flags to fptrunc, which will be fixed in a later patch.

john-brawn-arm committed:
[IR] Allow fast math flags on fptrunc and fpext

This consists of:
 * Make these instructions part of FPMathOperator.
 * Adjust bitcode/ir readers/writers to expect fast math flags on these
   instructions.
 * Make IRBuilder set the fast math flags on these instructions.
 * Adjust InstCombinerImpl::visitFPTrunc, as currently the fast math flags on
   all of the output instructions are set from the source operation. Instead we
   want each output instruction to get the fast math flags from their
   corresponding input instruction.
 * Update langref and release notes.
 * Update a bunch of tests.

john-brawn-arm committed:
Fix formatting.

tschuett commented on llvm/docs/LangRef.rst:
```diff
@@ -3730,10 +3730,10 @@ Fast-Math Flags
-floating-point transformations.
+:ref:`frem <i_frem>`, :ref:`fcmp <i_fcmp>`, :ref:`fptrunc <i_fptrunc>`,
+:ref:`fpext <i_fpext>`), and :ref:`phi <i_phi>`, :ref:`select <i_select>`, or
+:ref:`call <i_call>` instructions that return floating-point types may use the
```
The syntax of `fptrunc` and `fpext` forbids fast math flags. You may want to update the two sections.

jcranmer-intel replied:
Agree, `fptrunc` and `fpext` need updating to at least show where the fast-math flags may occur.

arsenm commented on llvm/test/Assembler/fast-math-flags.ll:
```diff
@@ -42,6 +42,14 @@ entry:
   %f = fneg float %x
 ; CHECK:  %f_vec = fneg <3 x float> %vec
   %f_vec = fneg <3 x float> %vec
+; CHECK: %g = fpext float %x to double
```
Needs bitcode compatibility test

arsenm commented on llvm/test/Assembler/fast-math-flags.ll:
```diff
@@ -148,6 +172,14 @@ entry:
+; CHECK: %g_vec = fptrunc nnan ninf <3 x float> %vec to <3 x half>
+  %g_vec = fptrunc ninf nnan <3 x float> %vec to <3 x half>
 ; CHECK:  ret float %e
   ret float %e
```
Test scalable vector case

john-brawn-arm committed:
Adjust based on review comments

Add fast math flags to fptrunc and fpext in langref
Add scalable vector test
Add test in Bitcode/compatibility.ll

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:
```diff
@@ -1875,13 +1873,17 @@ Instruction *InstCombinerImpl::visitFPTrunc(FPTruncInst &FPT) {
       Value *NarrowY = Builder.CreateFPTrunc(Y, Ty);
       Value *Sel = Builder.CreateSelect(Cond, NarrowY, X, "narrow.sel", Op);
+      if (auto *I = dyn_cast<Instruction>(Sel))
+        I->setFastMathFlags(Op->getFastMathFlags());
```
We already propagate FMF in `IRBuilder::CreateSelect`. BTW it is incorrect to preserve `ninf`: https://github.com/llvm/llvm-project/pull/116073#discussion_r1840831023

john-brawn-arm replied:
IRBuilder::CreateSelect uses the default fast math flags of the IRBuilder, which will be the flags of the fptrunc. So we need this to make sure the new select has the same flags as the old select.

dtcxzyw replied:
Sorry I got MDFrom and FMFSource mixed up.

nikic replied:
Is it possible to split off this change? The rest of the patch looks good to me.

john-brawn-arm replied:
Do you mean these specific changes for setting the fast math flag on select instructions, or also the change at the top of visitFPTrunc to set the fast math flags in the builder based on the fptrunc? Either way it causes failures in the llvm/test/Transforms/InstCombine/fpcast.ll and llvm/test/Transforms/InstCombine/fptrunc.ll tests. Without setting the fast math flags in the select the failures are because the select gets the fast math flags from the fptrunc. Without setting the flags in the builder the opposite happens and the fptrunc gets the fast math flags from the select.

arsenm replied:
I would split the instcombine change into a separate PR, and leave this one purely mechanical to add the flags

john-brawn-arm replied:
I don't think this is possible. Applying this patch first then later applying a patch to instcombine would cause test failures, as I mention in my comment above. Applying a patch to instcombine first doesn't work, because Instruction::getFastMathFlags fails an assert when the instruction isn't an FPMathOperator, which fptrunc wouldn't be at that point.

arsenm replied:
The failures are just lost flag regressions? Just take the regression and let it be fixed in the follow up

john-brawn-arm replied:
Done

john-brawn-arm commented:
Ping

john-brawn-arm committed:
Don't make any changes to InstCombineCasts, instead change the tests

john-brawn-arm committed:
Merge branch 'main' into fpcast_fast_math

nikic reviewed (approved):
LGTM

jcranmer-intel reviewed (approved):
I think there's some followup work needed to get this working in SelectionDAG as well, but that can live in a separate patch (especially as we need to do a followup pass in DAGCombine to check for flags on the fpext/fptrunc nodes).

john-brawn-arm commented:
Instcombine follow up : https://github.com/llvm/llvm-project/pull/118808
